File: sol/solution/0001-0100/0020/Solution.ex
defmodule Solution do
  @spec is_valid(s :: String.t) :: boolean
  def is_valid(s), do: match_brackets(s, [])
  defp match_brackets(<<?(, rest::binary>>, stack), do: match_brackets(rest, [?( | stack])
  defp match_brackets(<<?[, rest::binary>>, stack), do: match_brackets(rest, [?[ | stack])
  defp match_brackets(<<?{, rest::binary>>, stack), do: match_brackets(rest, [?{ | stack])
  defp match_brackets(<<?), rest::binary>>, [?( | stack]), do: match_brackets(rest, stack)
  defp match_brackets(<<?], rest::binary>>, [?[ | stack]), do: match_brackets(rest, stack)
  defp match_brackets(<<?}, rest::binary>>, [?{ | stack]), do: match_brackets(rest, stack)
  defp match_brackets("", []), do: true
  defp match_brackets(_, _), do: false
end

File: sol/solution/0001-0100/0020/Solution.erl
-spec is_valid(S :: unicode:unicode_binary()) -> boolean().
is_valid(S) ->
    Chars = unicode:characters_to_list(S),
    Pairs = #{ $) => $(, $] => $[, $} => ${ },
    check(Chars, [], Pairs).
check([], [], _Pairs) ->
    true;
check([], _Stack, _Pairs) ->
    false;
check([C | Rest], Stack, Pairs) ->
    case maps:get(C, Pairs, undefined) of
        undefined ->
            check(Rest, [C | Stack], Pairs);
        Open ->
            case Stack of
                [Open | Tail] -> check(Rest, Tail, Pairs);
                _ -> false
            end
    end.
